x-pack/plugins/ml/server/models/model_management/memory_usage.ts

```diff
@@ -0,0 +1,283 @@
+const AD_EXTRA_MEMORY = numeral('10MB').value();
+// @ts-expect-error numeral missing value
+const DFA_EXTRA_MEMORY = numeral('5MB').value();
+
+const NODE_FIELDS = ['attributes', 'name', 'roles'] as const;
+
+export type RequiredNodeFields = Pick<estypes.NodesInfoNodeInfo, typeof NODE_FIELDS[number]>;
+
+export class MemoryUsageService {
+  constructor(private readonly mlClient: MlClient) {}
+
+  public async getMemorySizes(itemType?: MlSavedObjectType, node?: string, showClosedJobs = false) {
+    let memories: MemoryUsageInfo[] = [];
+
+    switch (itemType) {
+      case 'anomaly-detector':
+        memories = await this.getADJobsSizes();
+        break;
+      case 'data-frame-analytics':
+        memories = await this.getDFAJobsSizes();
+        break;
+      case 'trained-model':
+        memories = await this.getTrainedModelsSizes();
+        break;
+      default:
+        memories = [
+          ...(await this.getADJobsSizes()),
+          ...(await this.getDFAJobsSizes()),
+          ...(await this.getTrainedModelsSizes()),
+        ];
+        break;
+    }
+    return memories.filter((m) => nodeFilter(m, node, showClosedJobs));
+  }
+
+  private async getADJobsSizes() {
+    const jobs = await this.mlClient.getJobStats();
+    return jobs.jobs.map(this.getADJobMemorySize);
+  }
+
+  private async getTrainedModelsSizes() {
+    const [models, stats] = await Promise.all([
+      this.mlClient.getTrainedModels(),
+      this.mlClient.getTrainedModelsStats(),
+    ]);
+    const statsMap = stats.trained_model_stats.reduce<Record<string, estypes.MlTrainedModelStats>>(
+      (acc, cur) => {
+        acc[cur.model_id] = cur;
+        return acc;
+      },
+      {}
+    );
+
+    return models.trained_model_configs.map((m) =>
+      this.getTrainedModelMemorySize(m, statsMap[m.model_id])
+    );
+  }
+
+  private async getDFAJobsSizes() {
+    const [jobs, jobsStats] = await Promise.all([
+      this.mlClient.getDataFrameAnalytics(),
+      this.mlClient.getDataFrameAnalyticsStats(),
+    ]);
+    const statsMap = jobsStats.data_frame_analytics.reduce<
+      Record<string, estypes.MlDataframeAnalytics>
+    >((acc, cur) => {
+      acc[cur.id] = cur;
+      return acc;
+    }, {});
+
+    return jobs.data_frame_analytics.map((j) => this.getDFAJobMemorySize(j, statsMap[j.id]));
+  }
+
+  private getADJobMemorySize(jobStats: JobStats): MemoryUsageInfo {
+    let memory = 0;
+    switch (jobStats.model_size_stats.assignment_memory_basis) {
+      case 'model_memory_limit':
+        memory = (jobStats.model_size_stats.model_bytes_memory_limit as number) ?? 0;
+        break;
+      case 'current_model_bytes':
+        memory = jobStats.model_size_stats.model_bytes as number;
+        break;
+      case 'peak_model_bytes':
+        memory = (jobStats.model_size_stats.peak_model_bytes as number) ?? 0;
+        break;
+    }
+
+    const size = memory + AD_EXTRA_MEMORY;
+    const text = numeral(size).format('0.000 b');
```

Comment:
shall we use a `fieldFormatter` plugin instead?

Reply:
This comment made me remember that the text version of the memory size isn't actually needed in the endpoint response. only the bytes value is used and is formatted when being displayed.

Removed in https://github.com/elastic/kibana/pull/149419/commits/882e0dd38a432a6b215884289b86edc84a35a7a0